# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/nsCaret.cpp
# Commit: caab023e5d2d
# Full Hash: caab023e5d2d6daee7993353cecc142dde2483fb
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2024-03-11 21:13:39
# Regressor Bug: 1860328
# File Overlap Count: 2
# Description:
#   Bug 1860328 - Fix a bug where caret misses when dragging and dropping r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D192377
# ==============================================================================

diff -r 09b0c2bfb164 -r caab023e5d2d layout/base/nsCaret.cpp
--- a/layout/base/nsCaret.cpp	Mon Mar 11 13:36:59 2024 +0000
+++ b/layout/base/nsCaret.cpp	Mon Mar 11 13:41:27 2024 +0000
@@ -398,6 +398,10 @@
 }
 
 void nsCaret::SchedulePaint(Selection* aSelection) {
+  if (mLastCaretFrame) {
+    mLastCaretFrame->MarkNeedsDisplayItemRebuild();
+  }
+
   Selection* selection;
   if (aSelection) {
     selection = aSelection;
@@ -409,14 +413,16 @@
   nsIFrame* frame = GetFrameAndOffset(selection, mOverrideContent,
                                       mOverrideOffset, &frameOffset);
   if (!frame) {
+    mLastCaretFrame = nullptr;
     return;
   }
 
   if (nsIFrame* cb = GetContainingBlockIfNeeded(frame)) {
-    cb->SchedulePaint();
-  } else {
-    frame->SchedulePaint();
+    frame = cb;
   }
+
+  mLastCaretFrame = frame;
+  frame->SchedulePaint();
 }
 
 void nsCaret::SetVisibilityDuringSelection(bool aVisibility) {
@@ -555,6 +561,9 @@
 NS_IMETHODIMP
 nsCaret::NotifySelectionChanged(Document*, Selection* aDomSel, int16_t aReason,
                                 int32_t aAmount) {
+  if (mLastCaretFrame) {
+    mLastCaretFrame->MarkNeedsDisplayItemRebuild();
+  }
   // Note that aDomSel, per the comment below may not be the same as our
   // selection, but that's OK since if that is the case, it wouldn't have
   // mattered what IsVisible() returns here, so we just opt for checking